// @cloudbase/framework-plugin-website/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = void 0;
/**
 * Tencent is pleased to support the open source community by making CloudBaseFramework - 云原生一体化部署工具 available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Please refer to license text included with this package for license details.
 */
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const lodash_merge_1 = __importDefault(require("lodash.merge"));
const framework_core_1 = require("@cloudbase/framework-core");
const static_builder_1 = require("@cloudbase/static-builder");
const zip_builder_1 = require("./zip-builder");
const toolbox_1 = require("@cloudbase/toolbox");
const DEFAULT_INPUTS = {
    outputPath: 'dist',
    cloudPath: '/',
    ignore: ['.git', '.github', 'node_modules', 'cloudbaserc.js'],
    commands: {
        install: 'npm install --prefer-offline --no-audit --progress=false',
    },
};
var PAYMODE;
(function (PAYMODE) {
    PAYMODE["POSTPAID"] = "postpaid";
    PAYMODE["PREPAYMENT"] = "prepayment";
})(PAYMODE || (PAYMODE = {}));
var ENV_CHANNEL;
(function (ENV_CHANNEL) {
    ENV_CHANNEL["LOWCODE"] = "low_code";
    ENV_CHANNEL["SCENEMODULE"] = "scene_module";
})(ENV_CHANNEL || (ENV_CHANNEL = {}));
class WebsitePlugin extends framework_core_1.Plugin {
    constructor(name, api, inputs) {
        super(name, api, inputs);
        this.name = name;
        this.api = api;
        this.inputs = inputs;
        this.resolvedInputs = resolveInputs(this.inputs);
        this.builder = new static_builder_1.StaticBuilder({
            projectPath: this.api.projectPath,
            copyRoot: path_1.default.join(this.api.projectPath, this.resolvedInputs.outputPath),
        });
        this.zipBuilder = new zip_builder_1.ZipBuilder({
            projectPath: this.api.projectPath,
        });
        this.buildOutput = {};
    }
    /**
     * 初始化
     */
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            this.api.logger.debug('WebsitePlugin: init', this.resolvedInputs);
            this.api.logger.info('Website 插件会自动开启静态网页托管能力，需要当前环境为 [按量计费] 模式');
            this.api.logger.info(`Website 插件会部署应用资源到当前静态托管的 ${this.resolvedInputs.cloudPath} 目录下`);
            if (this.resolvedInputs.httpPath) {
                this.api.logger.info(`Website 插件会配置触发路径为 ${this.resolvedInputs.httpPath} 的HTTP访问服务到静态托管`);
            }
            yield Promise.all([this.ensurePostPay(), this.fetchHostingInfo()]);
        });
    }
    /**
     * 编译为 SAM 模板
     */
    compile() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const uploadResults = yield this.upload();
            this.api.logger.debug('website uploadResults', uploadResults);
            const [website, staticConfig] = uploadResults;
            return {
                EnvType: ((_a = this.env) === null || _a === void 0 ? void 0 : _a.PayMode) === PAYMODE.PREPAYMENT ? 'PrePay' : 'PostPay',
                Resources: Object.assign({}, this.getStaticResourceSam('Website', '为开发者提供静态网页托管的能力，包括HTML、CSS、JavaScript、字体等常见资源。', website.codeUri, website.cloudPath), this.getStaticResourceSam('ConfigEnv', '配置文件', staticConfig.codeUri, staticConfig.cloudPath)),
                EntryPoint: [
                    {
                        Label: '网站入口',
                        EntryType: 'StaticStore',
                        HttpEntryPath: this.resolvedInputs.cloudPath,
                    },
                ],
            };
        });
    }
    getStaticResourceSam(name, description, codeUri, deployPath) {
        const httpPathConfig = this.resolvedInputs.httpPath && name === 'Website' ? { HttpPath: this.resolvedInputs.httpPath } : {};
        return {
            [name]: {
                Type: 'CloudBase::StaticStore',
                Properties: Object.assign({ Description: description ||
                        '为开发者提供静态网页托管的能力，静态资源（HTML、CSS、JavaScript、字体等）的分发由对象存储 COS 和拥有多个边缘网点的 CDN 提供支持', CodeUri: codeUri, DeployPath: deployPath }, httpPathConfig),
            },
        };
    }
    upload() {
        return __awaiter(this, void 0, void 0, function* () {
            const deployContent = [
                ...(this.buildOutput.static || []),
                ...(this.buildOutput.staticConfig || []),
            ];
            let zipFiles = (yield this.zipBuilder.build(deployContent.map((item, index) => {
                return {
                    name: item.name,
                    localPath: item.src,
                    zipFileName: `static-${index}.zip`,
                    ignore: this.resolvedInputs.ignore,
                };
            }))).zipFiles;
            this.api.logger.debug('website zipFiles', zipFiles);
            return Promise.all(deployContent.map((item, index) => __awaiter(this, void 0, void 0, function* () {
                let zipFile = zipFiles[index];
                let codeUris = (yield this.api.samManager.uploadFile([
                    {
                        fileType: 'STATIC',
                        fileName: zipFile.entry,
                        filePath: zipFile.source,
                    },
                ]));
                return Object.assign({}, item, {
                    codeUri: codeUris[0].codeUri,
                });
            })));
        });
    }
    /**
     * 删除资源
     */
    remove() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /**
     * 生成代码
     */
    genCode() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /**
     * 构建
     */
    build() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            // cloudPath 会影响 publicPath 和 baseRoute 等配置，需要处理
            this.api.logger.debug('WebsitePlugin: build', this.resolvedInputs);
            this.resolvedInputs.envVariables = Object.assign({
                TCB_SERVICE_DOMAIN: yield this.fetchServiceDomain(),
                TCB_ENV_ID: this.api.envId,
                TCB_REGION: process.env.CLOUDBASE_TCB_CLOUDAPI_REGION || (yield (0, toolbox_1.getRegion)()),
            }, this.resolvedInputs.envVariables);
            yield this.installPackage();
            const { cloudPath, buildCommand, envVariables, commands, } = this.resolvedInputs;
            const command = buildCommand || (commands === null || commands === void 0 ? void 0 : commands.build);
            if (command) {
                this.api.logger.info('running', command);
                yield this.api.spawnPromise(command, [], {
                    env: Object.assign({}, process.env, envVariables),
                });
            }
            const includes = [
                '**',
                ...this.resolvedInputs.ignore.map((ignore) => `!${ignore}`),
            ];
            this.buildOutput = yield this.builder.build(includes, {
                path: cloudPath,
                domain: (_a = this === null || this === void 0 ? void 0 : this.website) === null || _a === void 0 ? void 0 : _a.cdnDomain,
                config: envVariables,
            });
        });
    }
    /**
     * 部署
     */
    deploy() {
        return __awaiter(this, void 0, void 0, function* () {
            this.api.logger.debug('WebsitePlugin: deploy', this.resolvedInputs, this.buildOutput);
            this.api.logger.info(`${this.api.emoji('🚀')} 网站部署成功`);
            yield this.zipBuilder.clean();
            yield this.builder.clean();
        });
    }
    /**
     * 执行本地命令
     */
    run(params) {
        return __awaiter(this, void 0, void 0, function* () {
            this.api.logger.debug(`WebsitePlugin: run ${params === null || params === void 0 ? void 0 : params.runCommandKey}`);
            const { commands, envVariables } = this.resolvedInputs;
            const command = commands[params === null || params === void 0 ? void 0 : params.runCommandKey];
            if (!command)
                return;
            this.api.logger.info('running', command);
            return this.api.spawnPromise(command, [], {
                env: Object.assign({}, process.env, envVariables),
            });
        });
    }
    /**
     * 安装依赖
     */
    installPackage() {
        return __awaiter(this, void 0, void 0, function* () {
            const { installCommand, commands, envVariables } = this.resolvedInputs;
            const command = installCommand || (commands === null || commands === void 0 ? void 0 : commands.install);
            try {
                if (fs_1.default.statSync('package.json')) {
                    this.api.logger.info('running', command);
                    return this.api.spawnPromise(command, [], {
                        env: Object.assign({}, process.env, envVariables),
                    });
                }
            }
            catch (e) { }
        });
    }
    /**
     * 确保非BAAS环境开启了按量付费
     */
    ensurePostPay() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield this.api.cloudApi.tcbService.request('DescribeEnvs');
            let env = (_a = res === null || res === void 0 ? void 0 : res.EnvList) === null || _a === void 0 ? void 0 : _a[0];
            if (!env) {
                throw new Error(`当前账号下不存在 ${this.api.envId} 环境`);
            }
            if (env.EnvType !== 'baas' &&
                env.EnvType !== 'hosting' &&
                env.PayMode !== PAYMODE.POSTPAID &&
                env.EnvChannel !== ENV_CHANNEL.LOWCODE &&
                env.EnvChannel !== ENV_CHANNEL.SCENEMODULE) {
                throw new Error('网站托管当前只能部署到按量付费的环境下，请先在控制台切换计费方式');
            }
            this.env = env;
        });
    }
    /**
     * 查询静态托管信息
     */
    fetchHostingInfo() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const Hosting = (_a = this.api.resourceProviders) === null || _a === void 0 ? void 0 : _a.hosting;
            const envId = this.api.envId;
            if (!Hosting) {
                return;
            }
            let website;
            try {
                const hostingRes = yield Hosting.getHostingInfo({ envId });
                if (hostingRes.data.length) {
                    website = hostingRes.data[0];
                }
            }
            catch (e) {
                this.api.logger.debug(e);
            }
            this.website = website;
            return website;
        });
    }
    fetchServiceDomain() {
        return __awaiter(this, void 0, void 0, function* () {
            const serviceInfo = yield this.api.cloudApi.tcbUinService.request('DescribeCloudBaseGWService', {
                ServiceId: this.api.envId,
                EnableRegion: true,
            });
            return serviceInfo.DefaultDomain;
        });
    }
}
function resolveInputs(inputs) {
    return (0, lodash_merge_1.default)({}, DEFAULT_INPUTS, inputs);
}
exports.plugin = WebsitePlugin;
