// @cloudbase/framework-plugin-website/lib/zip-builder.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZipBuilder = void 0;
/**
 * Tencent is pleased to support the open source community by making CloudBaseFramework - 云原生一体化部署工具 available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Please refer to license text included with this package for license details.
 */
const path_1 = __importDefault(require("path"));
const archiver_1 = __importDefault(require("archiver"));
const fs_1 = __importDefault(require("fs"));
const framework_core_1 = require("@cloudbase/framework-core");
const toolbox_1 = require("@cloudbase/toolbox");
class ZipBuilder extends framework_core_1.Builder {
    constructor(options) {
        super(Object.assign({ type: 'zip' }, options));
    }
    build(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const zipFiles = yield Promise.all(options.map((option) => __awaiter(this, void 0, void 0, function* () {
                const localZipPath = path_1.default.join(this.distDir, option.zipFileName);
                if (!fs_1.default.existsSync(this.distDir)) {
                    (0, toolbox_1.mkdirSync)(this.distDir);
                }
                if (!fs_1.default.existsSync(option.localPath)) {
                    throw new Error(`目录或者文件 ${path_1.default.basename(option.localPath)} 不存在`);
                }
                const fileStats = fs_1.default.statSync(option.localPath);
                if (fileStats.isFile()) {
                    this.logger.debug('option.localPath', option.localPath, localZipPath);
                    yield this.zipFile(option.localPath, localZipPath);
                }
                else if (fileStats.isDirectory()) {
                    this.logger.debug('option.localPath', option.localPath, localZipPath);
                    yield this.zipDir(option.localPath, localZipPath, option.ignore);
                }
                return {
                    name: option.name,
                    options: {},
                    source: localZipPath,
                    entry: option.zipFileName,
                };
            })));
            return {
                zipFiles,
            };
        });
    }
    zipFile(src, dest) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                // create a file to stream archive data to.
                const output = fs_1.default.createWriteStream(dest);
                const archive = (0, archiver_1.default)('zip', {
                    zlib: { level: 9 }, // Sets the compression level.
                });
                output.on('close', () => {
                    resolve(void 0);
                });
                archive.on('error', reject);
                archive.file(src, {
                    name: path_1.default.basename(src),
                });
                archive.pipe(output);
                archive.finalize();
            });
        });
    }
    zipDir(src, dest, ignore) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                // create a file to stream archive data to.
                const output = fs_1.default.createWriteStream(dest);
                const archive = (0, archiver_1.default)('zip', {
                    zlib: { level: 9 }, // Sets the compression level.
                });
                output.on('close', () => {
                    resolve(void 0);
                });
                archive.on('error', reject);
                archive.glob('**/*', {
                    cwd: src,
                    ignore: ignore || [],
                    dot: true,
                }, {});
                archive.pipe(output);
                archive.finalize();
            });
        });
    }
}
exports.ZipBuilder = ZipBuilder;
