// @cloudbase/static-builder/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticBuilder = void 0;
/**
 * Tencent is pleased to support the open source community by making CloudBaseFramework - 云原生一体化部署工具 available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Please refer to license text included with this package for license details.
 */
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const cpy_1 = __importDefault(require("cpy"));
const framework_core_1 = require("@cloudbase/framework-core");
const toolbox_1 = require("@cloudbase/toolbox");
const CONFIG_FILE_NAME = 'cloudbaseenv.json';
const CONFIG_SCRIPT_FILE_NAME = '_init_tcb-env.js';
class StaticBuilder extends framework_core_1.Builder {
    constructor(options) {
        super(Object.assign({ type: 'static' }, options));
        this.copyRoot = options.copyRoot || this.projectDir;
    }
    build(includes, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const contentDistPath = path_1.default.join(this.distDir, 'content');
            const configDistPath = path_1.default.join(this.distDir, 'config');
            // build content
            yield (0, cpy_1.default)(includes, contentDistPath, {
                cwd: this.copyRoot,
                parents: true,
            });
            // build config
            yield this.buildConfig(configDistPath, options);
            return {
                static: [
                    {
                        src: contentDistPath,
                        cloudPath: options.path || '/',
                    },
                ],
                staticConfig: [
                    {
                        src: configDistPath,
                        cloudPath: '/',
                    },
                ],
                routes: [
                    {
                        path: options.path || '/',
                        targetType: 'static',
                        target: options.path || '/',
                    },
                ],
            };
        });
    }
    buildConfig(configDistPath, options) {
        return __awaiter(this, void 0, void 0, function* () {
            // 1. 读取旧配置
            let originConfig;
            if (options.domain && options.config) {
                const url = `https://${options.domain}/${CONFIG_FILE_NAME}`;
                const streamRes = yield (0, toolbox_1.fetchStream)(url, undefined, (0, toolbox_1.getProxy)());
                if ((streamRes === null || streamRes === void 0 ? void 0 : streamRes.status) == 200) {
                    originConfig = yield streamRes.json().catch(() => {
                        return {};
                    });
                }
                else {
                    originConfig = {};
                }
            }
            else {
                // 域名为空, 则静态托管还未开通
                originConfig = {};
            }
            // 2. 整合配置
            const resolvedConfig = Object.assign({}, originConfig, options.config);
            // 3. 写入新配置
            if (!fs_1.default.existsSync(configDistPath)) {
                (0, toolbox_1.mkdirSync)(configDistPath);
            }
            fs_1.default.writeFileSync(path_1.default.join(configDistPath, CONFIG_FILE_NAME), JSON.stringify(resolvedConfig));
            fs_1.default.writeFileSync(path_1.default.join(configDistPath, CONFIG_SCRIPT_FILE_NAME), `window._tcbEnv = ${JSON.stringify(resolvedConfig)};`);
        });
    }
}
exports.StaticBuilder = StaticBuilder;
